rust/operator-binary/src/listener.rs

```diff
@@ -0,0 +1,95 @@
+    utils::build_recommended_labels,
+    zk_controller::ZK_CONTROLLER_NAME,
+};
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Snafu, Debug)]
+pub enum Error {
+    #[snafu(display("Role {zk_role:?} is not defined in the ZooKeeperCluster spec"))]
+    InvalidRole {
+        source: crate::crd::Error,
+        zk_role: String,
+    },
+
+    #[snafu(display("object is missing metadata to build owner reference"))]
+    ObjectMissingMetadataForOwnerRef {
+        source: stackable_operator::builder::meta::Error,
+    },
+
+    #[snafu(display("failed to build recommended labels"))]
+    RecommendedLabels {
+        source: stackable_operator::builder::meta::Error,
+    },
+}
+
+pub fn build_role_listener(
+    zk: &v1alpha1::ZookeeperCluster,
+    zk_role: &ZookeeperRole,
+    resolved_product_image: &ResolvedProductImage,
+    zookeeper_security: &ZookeeperSecurity,
+) -> Result<listener::v1alpha1::Listener> {
+    let listener_name = role_listener_name(zk, zk_role);
+    let listener_class = &zk
+        .role(zk_role)
+        .with_context(|_| InvalidRoleSnafu {
+            zk_role: zk_role.to_string(),
+        })?
+        .role_config
+        .listener_class;
+
+    let listener = listener::v1alpha1::Listener {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(zk)
+            .name(listener_name)
+            .ownerreference_from_resource(zk, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            // Since we only make a listener for the role, which labels should we use?
+            // We can't use with_recommended_labels because it requires an ObjectLabels which
+            // in turn requires RoleGroup stuff)
+            // TODO (@NickLarsenNZ): Make separate functions for with_recommended_labels with/without rolegroups
+            // .with_labels(manual_labels).build()
+            .with_recommended_labels(build_recommended_labels(
+                zk,
+                ZK_CONTROLLER_NAME,
+                &resolved_product_image.app_version_label,
+                &zk_role.to_string(),
+                "global", // TODO (@NickLarsenNZ): update build_recommended_labels to have an optional role_group
+            ))
+            .context(RecommendedLabelsSnafu)?
+            .build(),
+        spec: listener::v1alpha1::ListenerSpec {
+            class_name: Some(listener_class.to_owned()),
+            ports: Some(listener_ports(zookeeper_security)),
+            ..listener::v1alpha1::ListenerSpec::default()
+        },
+        status: None,
+    };
+
+    Ok(listener)
+}
+
+// TODO (@NickLarsenNZ): This could be a method we can put on a Resource that takes a role_name
+pub fn role_listener_name(zk: &v1alpha1::ZookeeperCluster, zk_role: &ZookeeperRole) -> String {
+    // TODO (@NickLarsenNZ): Make a convention, do we use name_any() and allow empty string? or metadata.name.expect, or handle the error?
```

Comment:
We definitly do not want to expect here since its related to one CR.

Reply:
My point was that as far as the type system is concerned `name_any()` could return an empty String (due to `.unwrap_or_default()`). It isn't likely, and Kubernetes _promises_ to give a name since it requires one internally.

However, since we cannot guarantee this function is only ever called with data from K8s, we have to assume it can be used wrongly.

Therefore, I think we should error out.

But yet, not using expect.

How about:

```suggestion
    // TODO (@NickLarsenNZ): Make a convention, do we use name_any() and allow empty string? or handle the error (as unlikely as it would be)?
```